File: examples/nirfmx/evdo/csharp/VS2010/RFmxEvdoSlotPower/cs/RFmxEvdoSlotPower.cs
//Steps:
//1. Open a new RFmx Session.
//2. Configure Frequency Reference.
//3. Configure the basic signal properties (Center Frequency, Reference Level and External Attenuation).
//4. Configure Trigger Type and Trigger Parameters.
//5. Configure Uplink Spreading Parameters.
//6. Select SlotPower measurement.
//7. Configure Synchronization Mode and Interval
//8. Initiate the Measurement.
//9. Fetch SlotPower Measurement.
//10 Close the RFmx Seesion


 
using System;
using NationalInstruments.RFmx.InstrMX;
using NationalInstruments.RFmx.EvdoMX;

namespace NationalInstruments.Examples.RFmxEvdoSlotPower
{
    public class RFmxEvdoSlotPower
    {
        RFmxInstrMX instrSession;
        RFmxEvdoMX evdo;

        string resourceName;
        RFmxEvdoMXMeasurementTypes measurement;
        string frequencyReferenceSource;
        double frequencyReferenceFrequency;                                  
        double centerFrequency;                                              
        double externalAttenuation;                                          

        string digitalEdgeSource;
        RFmxEvdoMXDigitalEdgeTriggerEdge digitalEdge;
        double triggerDelay;                                                 
        double referenceLevel;                                              

        RFmxEvdoMXSlotPowerSynchronizationMode synchronizationMode;
        int measurementOffset;                                               
        int measurementLength;                                              
        long uplinkSpreadingIMask;
        long uplinkSpreadingQMask;

        bool enableAllTraces;
        bool enableTrigger;
        double timeout;

        double[] halfSlotPower = null;
        double[] halfSlotPowerDelta = null;

        public void Run()
        {
            try
            {
                InitializeVariables();
                InitializeInstr();
                ConfigureEvdo();
                RetrieveResults();
                PrintResults();
            }
            catch (Exception ex)
            {
                DisplayError(ex);
            }
            finally
            {
                /* Close session */
                CloseSession();
                Console.WriteLine("Press any key to exit");
                Console.ReadKey();
            }
        }

        private void InitializeVariables()
        {
            /* Initialize input variables */
            resourceName = "RFSA";
            measurement = RFmxEvdoMXMeasurementTypes.SlotPower;
            frequencyReferenceSource = RFmxInstrMXConstants.OnboardClock;
            frequencyReferenceFrequency = 10.0e+6;                             /* Hz */
            centerFrequency = 833.49e+6;                                       /* Hz */
            externalAttenuation = 0.000000;                                    /* dB */

            digitalEdgeSource = RFmxEvdoMXConstants.Pfi0;
            digitalEdge = RFmxEvdoMXDigitalEdgeTriggerEdge.Rising;
            triggerDelay = 0.000000;                                           /* seconds */
            referenceLevel = 0.000000;                                         /* dBm */

            synchronizationMode = RFmxEvdoMXSlotPowerSynchronizationMode.Slot;
            measurementOffset = 0;                                             /*slots*/
            measurementLength = 16;                                            /*slots*/
            uplinkSpreadingIMask = 0;
            uplinkSpreadingQMask = 0;

            enableAllTraces = true;
            enableTrigger = false;
            timeout = 10.0;
        }

        void InitializeInstr()
        {
            /* Create a new RFmx Session */
            instrSession = new RFmxInstrMX(resourceName, "");
        }

        void ConfigureEvdo()
        {
            evdo = instrSession.GetEvdoSignalConfiguration();
            instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequencyReferenceFrequency);
            evdo.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation);
            evdo.ConfigureDigitalEdgeTrigger("", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger);
            evdo.ConfigureUplinkSpreading("", uplinkSpreadingIMask, uplinkSpreadingQMask);
            evdo.SelectMeasurements("", measurement, enableAllTraces);
            evdo.SlotPower.Configuration.ConfigureSynchronizationModeAndInterval("", synchronizationMode, measurementOffset, measurementLength);
            evdo.Initiate("", "");
        }

        void RetrieveResults()
        {
            evdo.SlotPower.Results.FetchPowers("", timeout, ref halfSlotPower, ref halfSlotPowerDelta);
        }

        void PrintResults()
        {
            Console.WriteLine("------------Half Slot Powers------------\n");
            for (int i = 0; i < halfSlotPower.Length; i++)
            {
                Console.WriteLine("\nHalf Slot Number {0}", i);
                Console.WriteLine("Half Slot Power (dBm)                    : {0}", halfSlotPower[i]);
                Console.WriteLine("Half Slot Power Delta (dB)               : {0}", halfSlotPowerDelta[i]);
            }
        }

        void CloseSession()
        {
            if (evdo != null)
            {
                evdo.Dispose();
                evdo = null;
            }
            if (instrSession != null)
            {
                instrSession.Close();
                instrSession = null;
            }
        }

        static void DisplayError(Exception ex)
        {
            Console.WriteLine("ERROR:\n" + ex.GetType() + ": " + ex.Message);
        }

    }
}

File: examples/nirfmx/evdo/csharp/VS2010/RFmxEvdoSlotPower/vb/RFmxEvdoSlotPower.vb
'Steps:
'1. Open a new RFmx Session.
'2. Configure Frequency Reference.
'3. Configure the basic signal properties (Center Frequency, Reference Level and External Attenuation).
'4. Configure Trigger Type and Trigger Parameters.
'5. Configure Uplink Spreading Parameters.
'6. Select SlotPower measurement.
'7. Configure Synchronization Mode and Interval
'8. Initiate the Measurement.
'9. Fetch SlotPower Measurement.
'10 Close the RFmx Seesion



Imports NationalInstruments.RFmx.InstrMX
Imports NationalInstruments.RFmx.EvdoMX

Public Class RFmxEvdoSlotPower
	Private instrSession As RFmxInstrMX
	Private evdo As RFmxEvdoMX

	Private resourceName As String
	Private measurement As RFmxEvdoMXMeasurementTypes
	Private frequencyReferenceSource As String
	Private frequencyReferenceFrequency As Double
	Private centerFrequency As Double
	Private externalAttenuation As Double

	Private digitalEdgeSource As String
	Private digitalEdge As RFmxEvdoMXDigitalEdgeTriggerEdge
	Private triggerDelay As Double
	Private referenceLevel As Double

	Private synchronizationMode As RFmxEvdoMXSlotPowerSynchronizationMode
	Private measurementOffset As Integer
	Private measurementLength As Integer
	Private uplinkSpreadingIMask As Long
	Private uplinkSpreadingQMask As Long


	Private enableAllTraces As Boolean
	Private enableTrigger As Boolean
    Private timeout As Double

    Private halfSlotPower As Double() = Nothing
	Private halfSlotPowerDelta As Double() = Nothing

	Public Sub Run()
		Try
			InitializeVariables()
			InitializeInstr()
			ConfigureEvdo()
			RetrieveResults()
			PrintResults()
		Catch ex As Exception
			DisplayError(ex)
		Finally
			' Close session 

			CloseSession()
			Console.WriteLine("Press any key to exit")
			Console.ReadKey()
		End Try
	End Sub

	Private Sub InitializeVariables()
		' Initialize input variables 

		resourceName = "RFSA"
		measurement = RFmxEvdoMXMeasurementTypes.SlotPower
		frequencyReferenceSource = RFmxInstrMXConstants.OnboardClock
		frequencyReferenceFrequency = 10000000.0
		' Hz 
		centerFrequency = 833490000.0
		' Hz 
		externalAttenuation = 0.0
		' dB 

		digitalEdgeSource = RFmxEvdoMXConstants.Pfi0
		digitalEdge = RFmxEvdoMXDigitalEdgeTriggerEdge.Rising
		triggerDelay = 0.0
		' seconds 
		referenceLevel = 0.0
		' dBm 

		synchronizationMode = RFmxEvdoMXSlotPowerSynchronizationMode.Slot
		measurementOffset = 0
		'slots
		measurementLength = 16
		'slots
		uplinkSpreadingIMask = 0
		uplinkSpreadingQMask = 0

		enableAllTraces = True
		enableTrigger = False
        timeout = 10.0
    End Sub

	Private Sub InitializeInstr()
		' Create a new RFmx Session 

		instrSession = New RFmxInstrMX(resourceName, "")
	End Sub

	Private Sub ConfigureEvdo()
		evdo = instrSession.GetEvdoSignalConfiguration()
		instrSession.ConfigureFrequencyReference("", frequencyReferenceSource, frequencyReferenceFrequency)
		evdo.ConfigureRF("", centerFrequency, referenceLevel, externalAttenuation)
        evdo.ConfigureDigitalEdgeTrigger("", digitalEdgeSource, digitalEdge, triggerDelay, enableTrigger)
		evdo.ConfigureUplinkSpreading("", uplinkSpreadingIMask, uplinkSpreadingQMask)
		evdo.SelectMeasurements("", measurement, enableAllTraces)
		evdo.SlotPower.Configuration.ConfigureSynchronizationModeAndInterval("", synchronizationMode, measurementOffset, measurementLength)
		evdo.Initiate("", "")
	End Sub

	Private Sub RetrieveResults()
		evdo.SlotPower.Results.FetchPowers("", timeout, halfSlotPower, halfSlotPowerDelta)
	End Sub

	Private Sub PrintResults()
		Console.WriteLine("------------Half Slot Powers------------" & vbLf)
		For i As Integer = 0 To halfSlotPower.Length - 1
			Console.WriteLine(vbLf & "Half Slot Number {0}", i)
			Console.WriteLine("Half Slot Power (dBm)                    : {0}", halfSlotPower(i))
			Console.WriteLine("Half Slot Power Delta (dB)               : {0}", halfSlotPowerDelta(i))
		Next

	End Sub

	Private Sub CloseSession()
		If evdo IsNot Nothing Then
			evdo.Dispose()
			evdo = Nothing
		End If
		If instrSession IsNot Nothing Then
			instrSession.Close()
			instrSession = Nothing
		End If
	End Sub

	Private Shared Sub DisplayError(ex As Exception)
		Console.WriteLine("ERROR:" & vbLf & Convert.ToString(ex.[GetType]()) & ": " & ex.Message)
	End Sub

End Class
